In [12]:

!pip install -q -U torchao
!pip install transformers datasets peft trl accelerate

In [3]:
data = [
    {"instruction": "What is a variable in Python?", "response": "A variable is a name that refers to a value stored in memory. In Python you create one simply by assigning a value to a name, like x = 5, without declaring its type."},
    {"instruction": "What is a list in Python?", "response": "A list is an ordered, mutable collection of items, written with square brackets like [1, 2, 3]. Lists can hold items of different types and support operations like appending, removing, and slicing."},
    {"instruction": "What is a tuple in Python?", "response": "A tuple is an ordered collection of items similar to a list, but it is immutable, meaning its contents cannot be changed after creation. Tuples are written with parentheses, like (1, 2, 3)."},
    {"instruction": "What is a dictionary in Python?", "response": "A dictionary is a collection of key-value pairs, written with curly braces like {'name': 'Ali'}. It allows fast lookup of a value using its associated key instead of a numeric index."},
    {"instruction": "What is a set in Python?", "response": "A set is an unordered collection of unique items, written with curly braces like {1, 2, 3}. Sets automatically remove duplicates and support operations like union and intersection."},
    {"instruction": "What is the difference between a list and a tuple?", "response": "A list is mutable, so its elements can be changed, added, or removed after creation, while a tuple is immutable and cannot be modified once created. Tuples are generally faster and used when the data should not change."},
    {"instruction": "What is a function in Python?", "response": "A function is a reusable block of code defined with the def keyword that performs a specific task. It can accept input parameters and optionally return a value using the return statement."},
    {"instruction": "What is the difference between a parameter and an argument?", "response": "A parameter is the variable name listed in a function's definition, while an argument is the actual value passed to the function when it is called. For example, in def greet(name), name is a parameter, and greet('Ali') passes 'Ali' as the argument."},
    {"instruction": "What is a for loop in Python?", "response": "A for loop iterates over the items of a sequence, such as a list, tuple, or string, executing a block of code for each item. It is written as for item in sequence: followed by the indented code block."},
    {"instruction": "What is a while loop in Python?", "response": "A while loop repeatedly executes a block of code as long as a given condition remains true. It is useful when the number of iterations is not known in advance."},
    {"instruction": "What is the difference between a for loop and a while loop?", "response": "A for loop iterates over a known sequence or a fixed number of times, while a while loop continues running as long as a condition is true, without necessarily knowing the number of iterations in advance."},
    {"instruction": "What is indentation in Python and why does it matter?", "response": "Indentation is the whitespace at the start of a line that Python uses to define code blocks, such as the body of a function, loop, or if statement. Unlike many languages that use braces, Python relies entirely on consistent indentation to determine which lines belong together."},
    {"instruction": "What is a Python module?", "response": "A module is a single Python file containing functions, classes, or variables that can be imported and reused in other files. Modules help organize code into logical, reusable pieces."},
    {"instruction": "What is a Python package?", "response": "A package is a directory containing multiple related modules, along with an __init__.py file, that lets you organize a larger codebase into a structured, importable collection of modules."},
    {"instruction": "What does the import statement do?", "response": "The import statement loads a module or package into the current file so its functions, classes, or variables can be used. For example, import math gives access to functions like math.sqrt()."},
    {"instruction": "What is a class in Python?", "response": "A class is a blueprint for creating objects, defining the attributes and methods that its instances will have. Classes are defined with the class keyword and support object-oriented programming concepts like inheritance."},
    {"instruction": "What is an object in Python?", "response": "An object is an instance of a class, created by calling the class like a function, for example my_car = Car(). Each object has its own copy of the attributes defined in its class."},
    {"instruction": "What is the __init__ method used for?", "response": "The __init__ method is a special constructor method automatically called when a new object is created from a class. It is used to initialize the object's attributes with starting values."},
    {"instruction": "What is inheritance in Python?", "response": "Inheritance lets a class, called a child or subclass, reuse and extend the attributes and methods of another class, called a parent or superclass. It is written as class Child(Parent):, allowing code reuse and specialization."},
    {"instruction": "What is a list comprehension?", "response": "A list comprehension is a concise way to create a list by applying an expression to each item in an iterable, optionally with a condition. For example, [x*2 for x in range(5)] creates a list of doubled values."},
    {"instruction": "What is the range() function used for?", "response": "The range() function generates a sequence of numbers, commonly used to control the number of iterations in a for loop. For example, range(5) produces the numbers 0 through 4."},
    {"instruction": "What is slicing in Python?", "response": "Slicing extracts a portion of a sequence, such as a list or string, using the syntax sequence[start:stop:step]. For example, my_list[1:4] returns elements from index 1 up to, but not including, index 4."},
    {"instruction": "What is the difference between == and is in Python?", "response": "== checks whether two objects have equal values, while is checks whether two variables refer to the exact same object in memory. Two different objects can have equal values but not be the same object."},
    {"instruction": "What is a try-except block used for?", "response": "A try-except block handles errors gracefully by attempting to run code inside the try block and catching any exceptions that occur in the except block, preventing the program from crashing."},
    {"instruction": "What is an exception in Python?", "response": "An exception is an error that occurs during program execution, such as dividing by zero or accessing an undefined variable. Python raises exceptions as objects that can be caught and handled using try-except."},
    {"instruction": "What does the len() function do?", "response": "The len() function returns the number of items in a collection, such as the number of characters in a string, elements in a list, or key-value pairs in a dictionary."},
    {"instruction": "What is the difference between append() and extend() for lists?", "response": "append() adds a single item to the end of a list, even if that item is itself a list, while extend() adds each element of an iterable individually to the end of the list."},
    {"instruction": "What is a lambda function?", "response": "A lambda function is a small, anonymous function defined in a single line using the lambda keyword, without needing a full def statement. For example, lambda x: x * 2 creates a function that doubles its input."},
    {"instruction": "What does the return statement do?", "response": "The return statement exits a function and optionally sends a value back to the code that called it. If no value is specified, the function returns None by default."},
    {"instruction": "What is the difference between a local variable and a global variable?", "response": "A local variable is defined inside a function and can only be accessed within that function, while a global variable is defined outside any function and can be accessed throughout the entire file."},
    {"instruction": "What is string formatting with f-strings?", "response": "F-strings let you embed expressions directly inside string literals by prefixing the string with f and wrapping expressions in curly braces, like f'Hello {name}'. They are a concise and readable way to build formatted strings."},
    {"instruction": "What is the difference between a mutable and an immutable object?", "response": "A mutable object can be changed after it is created, such as a list or dictionary, while an immutable object cannot be changed, such as a string, integer, or tuple. Attempting to modify an immutable object actually creates a new object."},
    {"instruction": "What does the None value represent in Python?", "response": "None represents the absence of a value or a null value in Python. It is often used as a default return value or to indicate that a variable has not yet been assigned meaningful data."},
    {"instruction": "What is the purpose of the if __name__ == '__main__': block?", "response": "This block checks whether a Python file is being run directly or imported as a module into another file. Code inside it only runs when the file is executed directly, not when it is imported."},
    {"instruction": "What is a Python virtual environment?", "response": "A virtual environment is an isolated Python environment with its own installed packages, separate from the system-wide Python installation. It allows different projects to use different, potentially conflicting, package versions without interfering with each other."},
    {"instruction": "What does pip do in Python?", "response": "pip is Python's package installer, used to download and install third-party libraries from the Python Package Index. For example, pip install requests installs the requests library."},
]

In [13]:
from datasets import Dataset
formatted = [{"text": f"### Instruction:\n{d['instruction']}\n\n### Response:\n{d['response']}"} for d in data]
dataset = Dataset.from_list(formatted).train_test_split(test_size=0.1)

In [14]:
# Cell: load model
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

MODEL_NAME = "HuggingFaceTB/SmolLM2-135M-Instruct"
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
if tokenizer.pad_token is None:
    tokenizer.pad_token = tokenizer.eos_token
model = AutoModelForCausalLM.from_pretrained(MODEL_NAME, torch_dtype=torch.float16, device_map="auto")

Loading weights:   0%|          | 0/272 [00:00<?, ?it/s]

In [15]:
# Cell: ONE ask function only, define once, use everywhere
def ask(model, question):
    prompt = f"### Instruction:\n{question}\n\n### Response:\n"
    inputs = tokenizer(prompt, return_tensors="pt").to(model.device)
    out = model.generate(
        **inputs,
        max_new_tokens=80,
        min_new_tokens=15,
        do_sample=True,
        temperature=0.3,
        top_p=0.9,
        repetition_penalty=1.15,
        pad_token_id=tokenizer.pad_token_id,
    )
    return tokenizer.decode(out[0], skip_special_tokens=True)

In [16]:
# Cell: before results
questions = [
    "What is a list in Python?",
    "What is a function in Python?",
    "What is the difference between a list and a tuple?",
    "What is a dictionary in Python?",
    "What is inheritance in Python?",
]
before = {q: ask(model, q) for q in questions}
for q in questions:
    print("BEFORE:", before[q])
    print("-" * 60)

BEFORE: ### Instruction:
What is a list in Python?

### Response:
A list in Python is an ordered collection of items that can be of any data type, including strings, integers, floats, and other lists. Lists are mutable, meaning they can be modified after creation. They have the following properties:

1. **Order**: A list has a specific order for its elements. The first element will always come before the last element when you access it directly from
------------------------------------------------------------
BEFORE: ### Instruction:
What is a function in Python?

### Response:
A function in Python can be thought of as an `if` statement that takes one or more arguments and returns the result. It's essentially a block of code that performs some operation on those inputs, which are then passed to another part of the program for execution. Functions allow you to write reusable code by breaking it down into smaller, independent functions with clear names and docstrings (short descriptions)

In [17]:
# Cell: LoRA config
from peft import LoraConfig, get_peft_model, TaskType

lora_config = LoraConfig(
    r=8,
    lora_alpha=16,
    lora_dropout=0.05,
    target_modules=["q_proj", "v_proj"],
    task_type=TaskType.CAUSAL_LM,
)
peft_model = get_peft_model(model, lora_config)
peft_model.print_trainable_parameters()

trainable params: 460,800 || all params: 134,975,808 || trainable%: 0.3414


In [18]:
# Cell: train
from trl import SFTTrainer, SFTConfig

args = SFTConfig(
    output_dir="./output",
    num_train_epochs=3,
    per_device_train_batch_size=2,
    learning_rate=2e-4,
    dataset_text_field="text",
    max_length=256,
)
trainer = SFTTrainer(model=peft_model, args=args, train_dataset=dataset["train"], eval_dataset=dataset["test"], processing_class=tokenizer)
trainer.train()

Adding EOS to train dataset:   0%|          | 0/32 [00:00<?, ? examples/s]

Tokenizing train dataset:   0%|          | 0/32 [00:00<?, ? examples/s]

Building labels for train dataset:   0%|          | 0/32 [00:00<?, ? examples/s]

Truncating train dataset:   0%|          | 0/32 [00:00<?, ? examples/s]

Dropping fully masked examples from train dataset:   0%|          | 0/32 [00:00<?, ? examples/s]

Adding EOS to eval dataset:   0%|          | 0/4 [00:00<?, ? examples/s]

Tokenizing eval dataset:   0%|          | 0/4 [00:00<?, ? examples/s]

Building labels for eval dataset:   0%|          | 0/4 [00:00<?, ? examples/s]

Truncating eval dataset:   0%|          | 0/4 [00:00<?, ? examples/s]

Dropping fully masked examples from eval dataset:   0%|          | 0/4 [00:00<?, ? examples/s]

Step,Training Loss
10,2.070922
20,2.009770
30,1.988121
40,1.975349


TrainOutput(global_step=48, training_loss=1.978689173857371, metrics={'train_runtime': 37.6289, 'train_samples_per_second': 2.551, 'train_steps_per_second': 1.276, 'total_flos': 3916711480320.0, 'train_loss': 1.978689173857371, 'entropy': 1.7706583887338638, 'num_tokens': 5784.0, 'mean_token_accuracy': 0.6147028133273125, 'epoch': 3.0})

In [19]:
# Cell: save adapter
peft_model.save_pretrained("./my-lora-adapter")
tokenizer.save_pretrained("./my-lora-adapter")

('./my-lora-adapter/tokenizer_config.json',
 './my-lora-adapter/chat_template.jinja',
 './my-lora-adapter/tokenizer.json')

In [20]:
# Cell: after results, SAME ask() function, no redefinition
after = {q: ask(peft_model, q) for q in questions}
for q in questions:
    print("BEFORE:", before[q])
    print("AFTER: ", after[q])
    print("-" * 60)

[transformers] `use_cache=True` is incompatible with gradient checkpointing. Setting `use_cache=False`.
[transformers] Caching is incompatible with gradient checkpointing in LlamaDecoderLayer. Setting `past_key_values=None`.


BEFORE: ### Instruction:
What is a list in Python?

### Response:
A list in Python is an ordered collection of items that can be of any data type, including strings, integers, floats, and other lists. Lists are mutable, meaning they can be modified after creation. They have the following properties:

1. **Order**: A list has a specific order for its elements. The first element will always come before the last element when you access it directly from
AFTER:  ### Instruction:
What is a list in Python?

### Response:
A the and of_,. to have you it's1-2 that your" 3 with be keep inequality are from` or ` was

 we can "$640, their ensuring ensure learning for explore as```
( this while information by such not[ * =5. example what ( include, there give they provide on which I - the focus about understanding will function occur
------------------------------------------------------------
BEFORE: ### Instruction:
What is a function in Python?

### Response:
A function in Python can be thought o

# AI Engineering Lab: Fine-Tuning a Small Model with LoRA / PEFT

## Overview

This is a small, guided fine-tuning experiment. A small pre-trained model was adapted using LoRA / PEFT rather than training a model from scratch.

**Pipeline:** Base Model -> Dataset -> Tokenization -> LoRA Adapter -> SFT Training -> Fine-Tuned Behaviour -> Evaluation

## Stack Used

- Google Colab (T4 GPU)
- Hugging Face Transformers
- Hugging Face TRL (SFTTrainer)
- PEFT / LoRA
- Hugging Face Datasets

## Files in this repo

- `finetuning_lab.ipynb`, the Colab notebook with the full workflow
- `dataset.json`, the 35 hand-written instruction/response pairs
- `training_config.json`, the LoRA and training hyperparameters used
- `before_after_comparison.csv` / screenshots, before vs after outputs
- `README.md`, this file

## Conclusion

**1. What model did I use?**

I used `HuggingFaceTB/SmolLM2-135M-Instruct`, a small (~135 million parameter) instruction-tuned model. I chose it because it could be fine-tuned easily on Colab's free GPU without needing heavy hardware.

**2. What dataset did I create, and how many examples?**

I chose "Python basics" as my topic (questions and answers about variables, lists, functions, loops, classes, exceptions, and related concepts). I wrote 35 instruction-response pairs by hand, and split them roughly 90% (31) for training and 10% (4) for evaluation.

**3. Why did I use LoRA instead of full fine-tuning?**

Full fine-tuning would have updated all 135 million parameters of the model, which needs more GPU memory and more time. Instead, I used LoRA, which freezes the base model's weights and only trains small low-rank matrices. In my case, this meant only 460,800 trainable parameters, about 0.34% of the total. This made my training much faster, far less memory intensive, and gave me an adapter file that's only a few MB, while still meaningfully changing the model's behavior.

Training ran for 3 epochs (48 steps total). The training loss barely moved, going from about 2.07 to 1.97.

**4. What changed after fine-tuning?**

The results were the opposite of what I expected. The base model actually gave fairly coherent, on-topic answers before fine-tuning. For example, it answered "What is a list in Python?" with a mostly accurate description of Python lists (an ordered, mutable collection of items), and it answered "What is inheritance in Python?" with a reasonable explanation of how a class can inherit properties and behavior from another class.

After fine-tuning, the same questions produced garbled, incoherent text instead of clean answers, for example generating fragments like "A, the your. and of_1 to have that you it's2 this..." rather than a real explanation. So in this run, fine-tuning made the model's responses worse, not better.

**5. What problems did I notice?**

- Fine-tuning made the model's output noticeably worse, not better, which was the opposite of what I expected going in.
- The training loss barely decreased over the 48 training steps, which suggests the model was not learning the task well in the first place.
- My best explanation is that the dataset was very small (35 examples) and the number of training steps was low (48), and the learning rate (2e-4) was enough to disturb the base model's existing language fluency without teaching it a stable, useful pattern to replace it with. This looks like a mild form of catastrophic forgetting.
- This was a useful lesson: fine-tuning does not automatically improve a model. If the training setup is not well tuned, either more training data, more training steps, or a lower learning rate would likely be needed to get the model to genuinely improve instead of degrade.

**6. When would I use RAG instead of fine-tuning?**

I would use RAG when the information changes frequently or is too large to bake into the model's weights, such as a company's latest data or live documents, since RAG retrieves fresh information at query time without retraining. I would use fine-tuning when I want the model to consistently follow a particular style, tone, or narrow skill that should be permanently built into its behavior, rather than injecting or updating factual content. Given what I saw in this experiment, I would also add that fine-tuning needs a properly sized dataset and carefully chosen hyperparameters to actually help, otherwise RAG or prompt engineering can be the safer choice for a small project.

## Training Configuration

```json
{
  "base_model": "HuggingFaceTB/SmolLM2-135M-Instruct",
  "lora_config": {
    "r": 8,
    "lora_alpha": 16,
    "lora_dropout": 0.05,
    "target_modules": ["q_proj", "v_proj"],
    "bias": "none"
  },
  "training_args": {
    "num_train_epochs": 3,
    "per_device_train_batch_size": 2,
    "learning_rate": 0.0002,
    "max_length": 256
  },
  "results": {
    "trainable_params": 460800,
    "total_params": 134975808,
    "trainable_percent": 0.3414,
    "final_training_loss": 1.978689173857371
  }
}
```